### ISSUE TITLE
Finalize GitHub Sync

### ISSUE THREAD

**Issue by christad92** [OpenFn/lightning#2352]
- [x] [Make "Initiate Github Sync" take the latest snapshots at the time of initiation](https://github.com/OpenFn/lightning/issues/1827)

- [ ] [Support paths to job files in project.yaml files](https://github.com/OpenFn/kit/issues/285)
For users who work locally or for certain projects in their GitHub repos, users want to keep their jobs in different .js files and then have the files referenced in their project.yaml files. When OpenFn deploy is called, we need to make sure we get the content of the files and add them to the right section on the project.yaml. 
We should advise users to create a job.js file for every job such that for a workflow with 5 steps, we will need 5 jobs.js files.  that are referenced in the .yaml file.

See this example below , currently a job looks like this:

```
 Notify-CHW-upload-successful:
       name: Notify-CHW-upload-successful
       adaptor: @openfn/language-http@latest
       enabled: true
       credential:
       globals:
       body: |
         fn(state => state);
We'd expect users to set the body to the following, and for the deploy to still work. 

 Notify-CHW-upload-successful:
       name: Notify-CHW-upload-successful
       adaptor: @openfn/language-http@latest
       enabled: true
       credential:
       globals:
       body: Notify-CHW-upload-successful.js;

Notify-CHW-upload-successful.js should contain `fn(state => state)`;
```

- [ ] [Improve error handling and messaging for GitHub Integrations](https://github.com/OpenFn/lightning/issues/2156)
Since we are wrapping around GitHub APIs, we should be returning error messages that help users know how to diagnose and find solutions rather than returning generalized error messages like "Oops!...". The recommendation here will be that we don't process the error messages in any special manner but return it as `GitHub Error: {message}`. This way the user can try to search for solutions on their own or know what to communicate via the community. 

Nice to haves: 

- [ ] [Race condition when triggering sync from openfn while deploy actions are running](https://github.com/OpenFn/lightning/issues/2179) #2179
- [ ] [Unable to disconnect/reinstall github app from lightning after uninstalling directly](https://github.com/OpenFn/lightning/issues/2168) #2168
- [ ] [Generating project.yaml fails on certain projects](https://github.com/OpenFn/lightning/issues/1902) #1902
- [ ] [GitHub sync button available, but subscription doesn't allow github sync](https://github.com/OpenFn/thunderbolt/issues/188) thunderbolt/188

**Comment by stuartc** [comment]
Instead of putting a path directly in the `body` key (which would also be a string value), we use this pattern instead:

```yaml
  job-xyz:
    ...
    body:
      path: ../my-job.js
```

That way we can tell that the body is _not_ available directly under the key much more easily, and leaves room for other features.

Note that this path is **relative** to the project.yaml, and not the current working directory.

**Comment by DrCord** [comment]
The link to the issue `GitHub sync button available, but subscription doesn't allow github sync` issue above is broken. I am interested in the issue, but cannot find it when searching.

**Comment by christad92** [comment]
#2462

**Comment by taylordowns2000** [comment]
Noting that after #2707 and #2462 @midigofrank and @stuartc will reasses this epic to determine if/what still needs to be done.

**Comment by christad92** [comment]
###Goal(s) & KPI: 
#### Goals: 
1. Improve the developer experience for GH Sync 
2. Don’t overwrite unsynced changes
#### KPI: Improve Consultant NPS

#### Problem/Observation: 
Some issues have been reported concerning GitHub Sync recently that hamper the experience of consultants on the application: 
1. Users don’t know when a GitHub sync fails until they check GitHub actions
2. When deploying local changes to a GitHub profile, we trigger a Sync that overwrites all unsynced changes on app.openfn.org. 
3. We cannot trust users to always sync their changes to GitHub
4. Updating random non-openfn files automatically triggers a sync: 
5. Aleksa recently mistakenly overwrote a project with old changes because she was trying to update the readme file of a repo.

#### Constraints (time/budget/product/) 
N/A
Assumptions: 
N/A

#### What to be done: 
1. Alert a user when sync breaks
We want users to know when GitHub sync fails via an email and a log list of syncs showing on the version control page.  
2. Only OpenFn related files should trigger a sync. 
GitHub to OpenFn sync should not be triggered because I updated the readme of my GitHub repo
3. Allow users to set up auto-sync to GitHub for their projects. This makes it so that a project automatically syncs to GitHub every X minutes and reduces the risk of overwriting changes.
Allow users to set a prompt for GitHub Sync after X number of saves. If this is set, a user will be able to say after every 5 saves, 

```
I want to be able to Sync to GitHub so that I don’t lose my data.
Save 1….4 = normal Save 
Save 5 = trigger Save and Sync and loads modal for commit and reset save count.
```
4. Allow users to see the list of a project’s GitHub Syncs and their final state in the version control page. https://github.com/OpenFn/lightning/issues/2965

**Comment by taylordowns2000** [comment]
@theroinaochieng and @josephjclark , do we still need this issue open?

**Comment by josephjclark** [comment]
No, I don't think so.

Once the sandbox design doc is done I'd like to compare it to this issues just to remind ourselves. But I think the work is better represented in other issues now.
